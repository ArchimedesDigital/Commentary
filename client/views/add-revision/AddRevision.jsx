import RaisedButton from 'material-ui/RaisedButton';
import FlatButton from 'material-ui/FlatButton';
import FontIcon from 'material-ui/FontIcon';
import baseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

import {EditorState, ContentState, Modifier, RichUtils, convertToRaw} from 'draft-js';
import Editor from 'draft-js-plugins-editor';
import {stateToHTML} from 'draft-js-export-html';
import {stateFromHTML} from 'draft-js-import-html';

import createSingleLinePlugin from 'draft-js-single-line-plugin';
const singleLinePlugin = createSingleLinePlugin();

import createRichButtonsPlugin from 'draft-js-richbuttons-plugin';
const richButtonsPlugin = createRichButtonsPlugin();

const {
  // inline buttons
  ItalicButton, UnderlineButton,
  // block buttons
  ULButton
} = richButtonsPlugin;

AddRevision = React.createClass({

    childContextTypes: {
        muiTheme: React.PropTypes.object.isRequired,
    },

    getChildContext() {
        return {muiTheme: getMuiTheme(baseTheme)};
    },

    propTypes: {
        submitForm: React.PropTypes.func.isRequired,
        comment: React.PropTypes.object.isRequired,
    },

    getInitialState(){
        var revisionId = this.props.comment.revisions.length - 1;
        var revision = this.props.comment.revisions[revisionId]; // get newest revision
        return {
            revision: revision,

            titleEditorState: EditorState.createWithContent(ContentState.createFromText(revision.title)),
            textEditorState: EditorState.createWithContent(stateFromHTML(revision.text)),

            titleValue: '',
            textValue: '',
        };
    },

    mixins: [ReactMeteorData],

    getMeteorData() {

        return {

        };
    },

    onTitleChange(titleEditorState) {
        var titleHtml = stateToHTML(this.state.titleEditorState.getCurrentContent());
        var title = jQuery(titleHtml).text();
        this.setState({
            titleEditorState: titleEditorState,
            titleValue: title,
        });
    },

    onTextChange(textEditorState) {
        var textHtml = stateToHTML(this.state.textEditorState.getCurrentContent());
        this.setState({
            textEditorState: textEditorState,
            textValue: textHtml,
        });
    },

    handleSubmit(event) {
        // TODO: form validation
        event.preventDefault();

        // var error = this.validateStateForSubmit();

        // this.setState({
        //     snackbarOpen: error.errors,
        //     snackbarMessage: error.errorMessage,
        // });
        // if (!error.errors) {
            this.props.submitForm(this.state);
        // };
    },

    selectRevision(event) {
        var revision = this.props.comment.revisions[event.currentTarget.id];
        this.setState({
            revision: revision,
            titleEditorState: EditorState.createWithContent(ContentState.createFromText(revision.title)),
            textEditorState: EditorState.createWithContent(stateFromHTML(revision.text)),
        });
    },

    removeRevision() { // TODO: delete
        console.log('this.state.revision', this.state.revision);
        Meteor.call('comment.remove.revision', this.props.comment._id, this.state.revision);
    },

    // validateStateForSubmit() {
    //     var errors = false;
    //     var errorMessage = "Missing comment data:"
    //     if(this.state.titleValue === ""){
    //         errors = true;
    //         errorMessage += " title,";
    //     };
    //     if(this.state.textValue === "<p><br></p>"){
    //         errors = true;
    //         errorMessage += " comment text,";
    //     };
    //     if(this.props.selectedLineFrom === 0){
    //         errors = true;
    //         errorMessage += " no line selected,";
    //     };
    //     if(errors === true) {
    //         errorMessage.slice(0,-1);
    //         errorMessage += ".";
    //     };
    //     return {
    //         errors: errors,
    //         errorMessage: errorMessage,
    //     };
    // },

    render() {

        var that = this;

        return (
					<div className="comments lemma-panel-visible">
            <div className={'comment-outer'}>

                <article className="comment commentary-comment paper-shadow " style={{marginLeft: 0}}>

                    <div className="comment-upper">
                        <h1 className="add-comment-title">
                            <Editor
                                editorState={this.state.titleEditorState}
                                onChange={this.onTitleChange}
                                placeholder='Comment title...'
                                spellCheck={true}
                                stripPastedStyles={true}
                                plugins={[singleLinePlugin]}
                                blockRenderMap={singleLinePlugin.blockRenderMap}
                            />
                        </h1>
                        <div className="comment-keywords">
                            {this.props.comment.keywords.map(function(keyword, i){
                                return <RaisedButton
                                        key={i}
                                        className="comment-keyword paper-shadow"
                                        onClick={self.addSearchTerm}
                                        data-id={keyword._id}
                                        label={(keyword.title || keyword.wordpressId)}
                                    />

                             })}
                        </div>
                        {/*TODO: this.props.comment.keyideas*/}

                    </div>
                    <div className="comment-lower" style={{paddingTop: 20}}>
                        <ItalicButton/>
                        <UnderlineButton/>
                        <ULButton/>
                        <div className="add-comment-text">
                            <Editor
                                editorState={this.state.textEditorState}
                                onChange={this.onTextChange}
                                placeholder='Comment text...'
                                spellCheck={true}
                                stripPastedStyles={true}
                                plugins={[richButtonsPlugin]}
                            />
                        </div>

                        <div className="comment-reference" >
                            <h4>Secondary Source(s):</h4>
                            <p>
                                {this.props.comment.referenceLink ?
                                    <a href={this.props.comment.referenceLink} target="_blank" >
                                        {this.props.comment.reference}
                                    </a>
                                    :
                                    <span >
                                        {this.props.comment.reference}
                                    </span>
                                }
                            </p>
                        </div>

                        <div className="add-comment-button">
                            <RaisedButton
                                type="submit"
                                label="Add revision"
                                labelPosition="after"
                                onClick={this.handleSubmit}
                                icon={<FontIcon className="mdi mdi-plus" />}
                            />
                        </div>
                        {Roles.userIsInRole(Meteor.user(), ['developer']) ? /*TODO: delete*/
                            <div className="add-comment-button">
                                <RaisedButton
                                    type="submit"
                                    label="(developer only) Remove revision"
                                    labelPosition="after"
                                    onClick={this.removeRevision}
                                    icon={<FontIcon className="mdi mdi-minus" />}
                                />
                            </div>
                            :
                            ""
                        }

                    </div>

                    <div className="comment-revisions">
                        {this.props.comment.revisions.map(function(revision, i){
                            return <FlatButton
                                key={i}
                                id={i}
                                className="revision selected-revision"
                                onClick={that.selectRevision}
                                label={"Revision " + moment(revision.created).format('D MMMM YYYY')}
                                >

                            </FlatButton>

                        })}
                    </div>

                </article>

            </div>
					</div>

        );
    }
});
