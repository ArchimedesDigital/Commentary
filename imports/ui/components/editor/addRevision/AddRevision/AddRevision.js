import React from 'react';
import { Meteor } from 'meteor/meteor';
import { Roles } from 'meteor/alanning:roles';
import { Session } from 'meteor/session';
import { createContainer } from 'meteor/react-meteor-data';
import {
	FormGroup,
	ControlLabel,
} from 'react-bootstrap';
import cookie from 'react-cookie';
import RaisedButton from 'material-ui/RaisedButton';
import TextField from 'material-ui/TextField';
import FlatButton from 'material-ui/FlatButton';
import FontIcon from 'material-ui/FontIcon';
import IconButton from 'material-ui/IconButton';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import { Creatable } from 'react-select';
import Formsy from 'formsy-react';
import { FormsyText } from 'formsy-material-ui/lib';
import { EditorState, ContentState, convertFromHTML, convertFromRaw, convertToRaw } from 'draft-js';
import Editor from 'draft-js-plugins-editor';
import { stateToHTML } from 'draft-js-export-html';
import { stateFromHTML } from 'draft-js-import-html';
import createSingleLinePlugin from 'draft-js-single-line-plugin';
import { fromJS } from 'immutable';
import update from 'immutability-helper';
import { convertToHTML } from 'draft-convert';
import createMentionPlugin, { defaultSuggestionsFilter } from 'draft-js-mention-plugin';
import createInlineToolbarPlugin, { Separator } from 'draft-js-inline-toolbar-plugin';
import {
	ItalicButton,
	BoldButton,
	UnderlineButton,
	UnorderedListButton,
	OrderedListButton,
	BlockquoteButton,
} from 'draft-js-buttons';

// api
import Keywords from '/imports/api/collections/keywords';
import ReferenceWorks from '/imports/api/collections/referenceWorks';

// components
import { ListGroupDnD, creatListGroupItemDnD } from '/imports/ui/components/shared/ListDnD';
import LinkButton from '/imports/ui/components/editor/addComment/LinkButton';
import AddTagInput from '/imports/ui/components/editor/addComment/AddTagInput';

// lib:
import muiTheme from '/imports/lib/muiTheme';

// helpers:
import linkDecorator from '/imports/ui/components/editor/addComment/LinkButton/linkDecorator';


// Create toolbar plugin for editor
const singleLinePlugin = createSingleLinePlugin();
const inlineToolbarPlugin = createInlineToolbarPlugin({
	structure: [
		BoldButton,
		ItalicButton,
		UnderlineButton,
		Separator,
		UnorderedListButton,
		OrderedListButton,
		BlockquoteButton,
		LinkButton,
	]
});
const { InlineToolbar } = inlineToolbarPlugin;

// Keyword Mentions
const keywordMentionPlugin = createMentionPlugin();

// Comments Cross Reference Mentions
const commentsMentionPlugin = createMentionPlugin({
	mentionTrigger: '#',
});

const ListGroupItemDnD = creatListGroupItemDnD('referenceWorkBlocks');

function _getSuggestionsFromComments(comments) {
	const suggestions = [];

	// if there are comments:
	if (comments.length) {

		// loop through all comments
		// add suggestion for each comment
		comments.forEach((comment) => {

			// get the most recent revision
			const revision = comment.revisions[comment.revisions.length - 1];

			const suggestion = {
				// create suggestio name:
				name: `"${revision.title}" -`,

				// set link for suggestion
				link: `/commentary?_id=${comment._id}`,

				// set id for suggestion
				id: comment._id,
			};

			// loop through commenters and add them to suggestion name
			comment.commenters.forEach((commenter, i) => {
				if (i === 0) suggestion.name += ` ${commenter.name}`;
				else suggestion.name += `, ${commenter.name}`;
			});

			suggestions.push(suggestion);
		});
	}
	return suggestions;
}

const AddRevision = React.createClass({

	propTypes: {
		submitForm: React.PropTypes.func.isRequired,
		update: React.PropTypes.func.isRequired,
		comment: React.PropTypes.object.isRequired,
		tags: React.PropTypes.array,
		referenceWorkOptions: React.PropTypes.array,
		isTest: React.PropTypes.bool,
	},

	getInitialState() {
		const { comment } = this.props;
		const revisionId = comment.revisions.length - 1;
		const revision = comment.revisions[revisionId]; // get newest revision
		let revisionTitle = '';
		
		const tagsValue = [];
		if (comment.keywords) {
			comment.keywords.forEach((keyword) => {
				tagsValue.push({
					isSet: true,
					keyword,
					tagId: keyword._id,
					isMentionedInLemma: keyword.isMentionedInLemma,
				});
			});
		}

		if (revision && revision.title) {
			revisionTitle = revision.title;
		}

		return {
			revision,

			titleEditorState: EditorState.createWithContent(ContentState.createFromText(revisionTitle)),
			textEditorState: this._getRevisionEditorState(revision),

			titleValue: '',
			textValue: '',

			tagsValue,

			referenceWorks: comment.referenceWorks || [],
			keywordSuggestions: fromJS([]),
			commentsSuggestions: fromJS([]),
		};
	},

	childContextTypes: {
		muiTheme: React.PropTypes.object.isRequired,
	},

	getChildContext() {
		return { muiTheme: getMuiTheme(muiTheme) };
	},

	_enableButton() {
		this.setState({
			canSubmit: true,
		});
	},

	_disableButton() {
		this.setState({
			canSubmit: false,
		});
	},

	_getRevisionEditorState(revision) {
		if (revision.textRaw) {
			return EditorState.createWithContent(convertFromRaw(revision.textRaw), linkDecorator);
		} else if (revision.text) {
			const blocksFromHTML = convertFromHTML(revision.text);
			return EditorState.createWithContent(
				ContentState.createFromBlockArray(
					blocksFromHTML.contentBlocks,
					blocksFromHTML.entityMap
				),
				linkDecorator
			);
		}
		console.error('missing filed text or textRaw in revision');
	},

	onTitleChange(titleEditorState) {
		const titleHtml = stateToHTML(this.state.titleEditorState.getCurrentContent());
		const title = jQuery(titleHtml).text();
		this.setState({
			titleEditorState,
			titleValue: title,
		});
	},

	onTextChange(textEditorState) {
		const newTextEditorState = EditorState.set(textEditorState, {decorator: linkDecorator});

		this.setState({
			textEditorState: newTextEditorState,
		});
	},

	onReferenceWorksValueChange(referenceWork) {
		const referenceWorks = this.state.referenceWorks;
		referenceWorks[referenceWork.i].referenceWorkId = referenceWork.value;

		this.setState({
			referenceWorks,
		});
	},

	_onKeywordSearchChange({ value }) {
		const keywordSuggestions = [];
		const keywords = this.props.tags;
		keywords.forEach((keyword) => {
			keywordSuggestions.push({
				name: keyword.label,
				link: `/keywords/${keyword.slug}`,
			});
		});

		this.setState({
			keywordSuggestions: defaultSuggestionsFilter(value, fromJS(keywordSuggestions)),
		});
	},

	_onCommentsSearchChange({ value }) {
		// use Meteor call method, as comments are not available on clint app
		Meteor.call('comments.getSuggestions', value, (err, res) => {
			// handle error:
			if (err) throw new Meteor.Error(err);

			// handle response:
			const commentsSuggestions = _getSuggestionsFromComments(res);

			this.setState({
				commentsSuggestions: fromJS(commentsSuggestions),
			});
		});

	},

	handleSubmit() {
		const { textEditorState } = this.state;

		// TODO: form validation
		// TODO: Migrate to formsy components

		// create html from textEditorState's content
		const textHtml = convertToHTML({

			// performe necessary html transformations:
			entityToHTML: (entity, originalText) => {

				// handle LINK
				if (entity.type === 'LINK') {
					return <a href={entity.data.link} target="_blank" rel="noopener noreferrer">{originalText}</a>;
				}

				// handle keyword mentions
				if (entity.type === 'mention') {
					return <a className="keyword-gloss" data-link={Utils.getEntityData(entity, 'link')}>{originalText}</a>;
				}

				// handle hashtag / commets cross reference mentions
				if (entity.type === '#mention') {
					return <a className="comment-cross-ref" href={Utils.getEntityData(entity, 'link')}><div dangerouslySetInnerHTML={{ __html: originalText }} /></a>;
				}
			},
		})(textEditorState.getCurrentContent());

		const textRaw = convertToRaw(textEditorState.getCurrentContent());

		this.props.submitForm(this.state, textHtml, textRaw);
	},

	handleUpdate() {
		const data = this.refs.form.getModel(); // eslint-disable-line
		let key;

		for (key in data) { // eslint-disable-line
			const params = key.split('_');
			params[0] = parseInt(params[0], 10);
			this.state.referenceWorks[params[0]][params[1]] = data[key];
		}
		this.props.update(this.state);
	},

	removeComment() {
		const authToken = cookie.load('loginToken');

		Meteor.call('comment.delete', authToken, this.props.comment._id, (err) => {
			if (err) {
				console.error(err);
				return false;
			}

			FlowRouter.go('/commentary');
		});
	},

	selectRevision(event) {
		const revision = this.props.comment.revisions[event.currentTarget.id];
		this.setState({
			revision,
			titleEditorState: EditorState.createWithContent(ContentState.createFromText(revision.title)),
			textEditorState: EditorState.createWithContent(stateFromHTML(revision.text), linkDecorator),
		});
	},

	removeRevision() {
		const self = this;
		Meteor.call('comment.remove.revision', this.props.comment._id, this.state.revision, (err) => {
			if (err) {
				throw new Meteor.Error('Error removing revision');
			}

			FlowRouter.go(`/commentary/${self.props.comment._id}/edit`);
		});
	},

	addReferenceWorkBlock() {
		this.state.referenceWorks.push({ referenceWorkId: '0' });
		this.setState({
			referenceWorks: this.state.referenceWorks,
		});
	},

	removeReferenceWorkBlock(i) {
		this.setState({
			referenceWorks: update(this.state.referenceWorks, { $splice: [[i, 1]] }),
		});
	},

	moveReferenceWorkBlock(dragIndex, hoverIndex) {
		const { introBlocks } = this.state;
		const dragIntroBlock = introBlocks[dragIndex];

		this.setState(update(this.state, {
			referenceWorks: {
				$splice: [
					[dragIndex, 1],
					[hoverIndex, 0, dragIntroBlock],
				],
			},
		}));
	},

	addTagBlock() {
		this.state.tagsValue.push({
			tagId: Random.id(),
			isMentionedInLemma: true,
			isSet: false,
		});
		this.setState({
			tagsValue: this.state.tagsValue,
		});
	},

	onTagValueChange(tag) {
		const tagsValue = this.state.tagsValue;

		tagsValue[tag.i].tagId = tag.value;
		tagsValue[tag.i].keyword = Keywords.findOne({_id: tag.value});
		tagsValue[tag.i].isSet = true;

		this.setState({
			tagsValue,
		});
	},

	onIsMentionedInLemmaChange(tag, i) {
		const tagsValue = this.state.tagsValue;

		tagsValue[i].isMentionedInLemma = !tag.isMentionedInLemma;
		
		this.setState({
			tagsValue,
		});
	},

	render() {
		const self = this;
		const { comment, isTest } = this.props;
		const { revision, titleEditorState, referenceWorks, textEditorState, tagsValue } = this.state;
		const { referenceWorkOptions, tags } = this.props;

		if (isTest) {
			return null;
		}

		return (
			<div className="comments lemma-panel-visible">
				<div className="comment-outer">

					<Formsy.Form
						ref="form" // eslint-disable-line
						onValid={this._enableButton}
						onInvalid={this._disableButton}
						onValidSubmit={this.handleSubmit}
					>
						<article className="comment commentary-comment paper-shadow " style={{ marginLeft: 0 }}>

							<div className="comment-upper">
								<div className="comment-action-buttons">
									<div className="comment-upper-action-button view-in-commentary">
										<FlatButton
											className="go-to-commentary-link"
											onClick={() => {
												FlowRouter.go('/commentary/', {}, {_id: comment._id});
											}}
											style={{
												border: '1px solid #ddd',
												maxHeight: 'none',
												fontSize: '12px',
												height: 'auto',
											}}
											label="View in Commentary"
										/>
									</div>
									<div className="comment-upper-action-button">
										<FlatButton
											label="Remove Comment"
											labelPosition="after"
											onClick={this.removeComment}
											style={{
												border: '1px solid #ddd',
												maxHeight: 'none',
												fontSize: '12px',
												height: 'auto',
											}}
										/>
									</div>
								</div>
								<h1 className="add-comment-title">
									{!isTest ?
										<Editor
											editorState={titleEditorState}
											onChange={this.onTitleChange}
											placeholder="Comment title..."
											spellCheck
											stripPastedStyles
											plugins={[singleLinePlugin]}
											blockRenderMap={singleLinePlugin.blockRenderMap}
										/>
									: ''}
								</h1>

								<AddTagInput
									tagsValue={tagsValue}
									tags={tags}
									addTagBlock={this.addTagBlock}
									onTagValueChange={this.onTagValueChange}
									onIsMentionedInLemmaChange={this.onIsMentionedInLemmaChange}
								/>

							</div>
							<div className="comment-lower clearfix" style={{ paddingTop: 20 }}>
								<Editor
									editorState={textEditorState}
									onChange={this.onTextChange}
									placeholder="Comment text..."
									spellCheck
									plugins={[commentsMentionPlugin, keywordMentionPlugin, inlineToolbarPlugin]}
									ref={(element) => { this.editor = element; }}
								/>

								{/* mentions suggestions for keywords */}
								<keywordMentionPlugin.MentionSuggestions
									onSearchChange={this._onKeywordSearchChange}
									suggestions={this.state.keywordSuggestions}
								/>

								{/* mentions suggestions for comments cross reference */}
								<commentsMentionPlugin.MentionSuggestions
									onSearchChange={this._onCommentsSearchChange}
									suggestions={this.state.commentsSuggestions}
								/>

								<div className="comment-reference">
									<h4>Secondary Source(s):</h4>
									<FormGroup
										controlId="referenceWorks"
										className="form-group--referenceWorks"
									>
										<ListGroupDnD>
											{/*
												DnD: add the ListGroupItemDnD component
												IMPORTANT:
												"key" prop must not be taken from the map function - has to be unique like _id
												value passed to the "key" prop must not be then edited in a FormControl component
													- will cause errors
												"index" - pass the map functions index variable here
											*/}
											{referenceWorks.map((referenceWork, i) => {
												const _referenceWorkOptions = [];
												referenceWorkOptions.forEach((rW) => {
													_referenceWorkOptions.push({
														value: rW.value,
														label: rW.label,
														slug: rW.slug,
														i,
													});
												});

												return (
													<ListGroupItemDnD
														key={referenceWork.referenceWorkId}
														index={i}
														className="form-subitem form-subitem--referenceWork"
														moveListGroupItem={this.moveReferenceWorkBlock}
													>
														<div
															className="reference-work-item"
														>
															<div
																className="remove-reference-work-item"
																onClick={this.removeReferenceWorkBlock.bind(this, i)}
															>
																<IconButton
																	iconClassName="mdi mdi-close"
																	style={{
																		padding: '0',
																		width: '32px',
																		height: '32px',
																		borderRadius: '100%',
																		border: '1px solid #eee',
																		color: '#666',
																		margin: '0 auto',
																		background: '#f6f6f6',
																	}}
																/>
															</div>
															<Creatable
																name="referenceWorks"
																id="referenceWorks"
																required={false}
																options={_referenceWorkOptions}
																value={this.state.referenceWorks[i].referenceWorkId}
																// onChange={this.onReferenceWorksValueChange.bind(this, referenceWork, i)}
																onChange={this.onReferenceWorksValueChange}
																placeholder="Reference Work . . ."
															/>
															<FormGroup>
																<ControlLabel>Section Number: </ControlLabel>
																<FormsyText
																	name={`${i}_section`}
																	defaultValue={referenceWork.section}
																/>
															</FormGroup>
															<FormGroup>
																<ControlLabel>Chapter Number: </ControlLabel>
																<FormsyText
																	name={`${i}_chapter`}
																	defaultValue={referenceWork.chapter}
																/>
															</FormGroup>
															<FormGroup>
																<ControlLabel>Translation Number: </ControlLabel>
																<FormsyText
																	name={`${i}_translation`}
																	defaultValue={referenceWork.translation}
																/>
															</FormGroup>
															<FormGroup>
																<ControlLabel>Note Number: </ControlLabel>
																<FormsyText
																	name={`${i}_note`}
																	defaultValue={referenceWork.note}
																/>
															</FormGroup>
														</div>
													</ListGroupItemDnD>
												);
											})}
										</ListGroupDnD>
										<RaisedButton
											label="Add Reference Work"
											onClick={this.addReferenceWorkBlock}
										/>
									</FormGroup>
								</div>


								<div className="comment-edit-action-button">
									<RaisedButton
										type="submit"
										label="Add revision"
										labelPosition="after"
										icon={<FontIcon className="mdi mdi-plus" />}
									/>
								</div>
								{(
									Roles.userIsInRole(Meteor.user(), ['editor', 'admin'])
									&& comment.revisions.length > 1
								) ?
									<div className="comment-edit-action-button comment-edit-action-button--remove">
										<RaisedButton
											label="Remove revision"
											labelPosition="after"
											onClick={this.removeRevision}
											icon={<FontIcon className="mdi mdi-minus" />}
										/>
									</div>
								: '' }
								<div className="comment-edit-action-button">
									<RaisedButton
										label="Update without adding Revision"
										labelPosition="after"
										icon={<FontIcon className="mdi mdi-plus" />}
										onClick={this.handleUpdate}
									/>
								</div>
							</div>


							<div className="comment-revisions">
								{comment.revisions.map((_revision, i) => (
									<FlatButton
										key={i}
										id={i}
										className={`revision ${revision._id === _revision._id ? 'selected-revision' : ''}`}
										onClick={self.selectRevision}
										label={`Revision ${moment(revision.created).format('D MMMM YYYY')}`}
									/>
								))}
							</div>
						</article>
					</Formsy.Form>
				</div>
				<div className="inline-toolbar-wrap">
					<InlineToolbar />
				</div>
			</div>
		);
	},
});

const AddRevisionContainer = createContainer(({ comment }) => {

	Meteor.subscribe('keywords.all', {tenantId: Session.get('tenantId')});

	const tags = Keywords.find().fetch();

	Meteor.subscribe('referenceWorks', Session.get('tenantId'));
	const referenceWorks = ReferenceWorks.find().fetch();
	const referenceWorkOptions = [];
	referenceWorks.forEach((referenceWork) => {
		if (!referenceWorkOptions.some((val) => (
			referenceWork.slug === val.slug
		))) {
			referenceWorkOptions.push({
				value: referenceWork._id,
				label: referenceWork.title,
				slug: referenceWork.slug,
			});
		}
	});

	return {
		tags,
		referenceWorkOptions,
	};
}, AddRevision);

export default AddRevisionContainer;
