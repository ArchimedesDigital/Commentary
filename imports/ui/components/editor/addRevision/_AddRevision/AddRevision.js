import React, {Component} from 'react';
import PropTypes from 'prop-types';
import { Meteor } from 'meteor/meteor';
import { Roles } from 'meteor/alanning:roles';
import { Session } from 'meteor/session';
import { createContainer } from 'meteor/react-meteor-data';
import { Field, reduxForm } from 'redux-form';
import cookie from 'react-cookie';
import RaisedButton from 'material-ui/RaisedButton';
import FlatButton from 'material-ui/FlatButton';
import IconButton from 'material-ui/IconButton';
import Utils from '/imports/lib/utils';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import { compose } from 'react-apollo';

// graphql
import { referenceWorksQuery } from '/imports/graphql/methods/referenceWorks';

// models
import Keywords from '/imports/models/keywords';
import ReferenceWorks from '/imports/models/referenceWorks';

// lib:
import muiTheme from '/imports/lib/muiTheme';

// components
import { ListGroupDnD, createListGroupItemDnD } from '/imports/ui/components/shared/ListDnD';
import CommentRevisionSelect from '/imports/ui/components/commentary/comments/CommentRevisionSelect';
import LinkButton from '../../addComment/LinkButton';
import TagsInput from '../../addComment/TagsInput';
import TitleInput from '../../addComment/TitleInput';
import ReferenceWorksInput from '../../addComment/ReferenceWorksInput';
import CommentContentInput from '../../addComment/CommentContentInput';
import CommentActionButtons from '../CommentActionButtons';
import AddRevisionButton from '../AddRevisionButton';
import RemoveRevisionButton from '../RemoveRevisionButton';
import UpdateRevisionButton from '../UpdateRevisionButton';




class AddRevision extends React.Component {

	static propTypes = {
		submitForm: PropTypes.func.isRequired,
		update: PropTypes.func.isRequired,
		comment: PropTypes.object.isRequired,
		tags: PropTypes.array,
		referenceWorkOptions: PropTypes.array,
		filters: PropTypes.array,
	}

	constructor(props) {
		super(props);

		this.state = {
			selectedRevisionIndex: null,
		};
	}

	getChildContext() {
		return { muiTheme: getMuiTheme(muiTheme) };
	}

	_enableButton() {
		this.setState({
			canSubmit: true,
		});
	}

	_disableButton() {
		this.setState({
			canSubmit: false,
		});
	}

	getRevisionIndex() {
		const { comment, filters } = this.props;
		let selectedRevisionIndex = this.state.selectedRevisionIndex;
		if (selectedRevisionIndex === null) {
			selectedRevisionIndex = 0;
		}
		return selectedRevisionIndex;
	}

	selectRevision(event) {
		this.setState({
			selectedRevisionIndex: parseInt(event.currentTarget.id, 10),
		});
	}

	handleSubmit() {
		const { textEditorState } = this.state;

		// create html from textEditorState's content
		const textHtml = Utils.getHtmlFromContext(textEditorState.getCurrentContent());

		const textRaw = convertToRaw(textEditorState.getCurrentContent());

		this.props.submitForm(this.state, textHtml, textRaw);
	}

	render() {
		const { comment } = this.props;
		const { referenceWorkOptions, tags } = this.props;

		const selectedRevisionIndex = this.getRevisionIndex();
		const selectedRevision = comment.revisions[selectedRevisionIndex];

		return (
			<form onSubmit={this.handleSubmit} >
				<div className="comments lemma-panel-visible">
					<div className="comment-outer">
						<article className="comment commentary-comment paper-shadow">
							<div className="comment-upper">
								<CommentActionButtons
									commentId={comment._id}
								/>
								<TitleInput
									placeholder="Comment title . . ."
								/>
								<TagsInput />
							</div>

							<div className="comment-lower clearfix">
								<CommentContentInput />
								<ReferenceWorksInput />
								<AddRevisionButton
									commentId={comment._id}
								/>
								<RemoveRevisionButton
									commentId={comment._id}
								/>
								<UpdateRevisionButton
									commentId={comment._id}
								/>
							</div>

							<div className="comment-revisions-outer">
								<CommentRevisionSelect
									commentId={comment._id}
									revisions={comment.revisions}
									comment={comment}
									selectedRevisionIndex={selectedRevisionIndex}
									selectRevision={this.selectRevision}
								/>
							</div>
						</article>
					</div>
				</div>
			</form>
		);
	}
}


const AddRevisionContainer = createContainer((props) => {

	const { comment } = props;
	const tenantId = Session.get('tenantId');
	Meteor.subscribe('keywords.all', {tenantId: tenantId});

	const tags = Keywords.find().fetch();

	if (tenantId) {
		props.referenceWorksQuery.refetch({
			tenantId: tenantId
		});
	}
	const referenceWorks = props.referenceWorksQuery.loading ? [] : props.referenceWorksQuery.referenceWorks;
	const referenceWorkOptions = [];
	referenceWorks.forEach((referenceWork) => {
		if (!referenceWorkOptions.some(val => (
			referenceWork.slug === val.slug
		))) {
			referenceWorkOptions.push({
				value: referenceWork._id,
				label: referenceWork.title,
				slug: referenceWork.slug,
			});
		}
	});

	return {
		tags,
		referenceWorkOptions,
	};
}, AddRevision);


AddRevision.childContextTypes = {
	muiTheme: PropTypes.object.isRequired,
};

/*
 * Make the redux form
 */
const AddRevisionForm = reduxForm({
	form: 'addRevision'
})(AddRevisionContainer);


export default compose(referenceWorksQuery)(AddRevisionForm);
