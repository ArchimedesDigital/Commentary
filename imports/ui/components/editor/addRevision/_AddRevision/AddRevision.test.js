import React from 'react';
import renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';

// component
import AddRevision from './AddRevision';

describe('AddRevision', () => {
	it('renders correctly', () => {

		const tree = renderer
			.create(
				<MuiThemeProvider>
					<AddRevision
						submitForm={() => {}}
						update={() => {}}
						comment={{
							revisions: [{
								title: 'testComment',
								text: 'test comment text',
							}],
						}}
					/>
				</MuiThemeProvider>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

// TODO Fix TypeError: Cannot read property 'split' of undefined (draft.js?)
