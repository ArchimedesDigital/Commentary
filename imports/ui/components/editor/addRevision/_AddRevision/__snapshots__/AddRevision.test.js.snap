// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRevision renders correctly 1`] = `
<div
  className="comments lemma-panel-visible"
/>
`;
